### ISSUE TITLE
Cannot run /bin/mpiexec

### ISSUE THREAD

**Issue by nfurmento** [open-mpi/ompi#12071]
### What version of Open MPI are you using? (e.g., v3.0.5, v4.0.2, git branch name and hash, etc.)
```
$ mpiexec -V
mpiexec (OpenRTE) 4.1.4
```

### Describe how Open MPI was installed (e.g., from a source/distribution tarball, from a git clone, from an operating system distribution package, etc.)
```
$ apt-cache policy openmpi-bin
openmpi-bin:
  Installé : 4.1.4-3+b1
  Candidat : 4.1.4-3+b1
 Table de version :
 *** 4.1.4-3+b1 990
        990 http://ftp.fr.debian.org/debian bookworm/main amd64 Packages
        100 /var/lib/dpkg/status
```

### Please describe the system on which you are running

```
$ uname -a
Linux vesubie 6.1.0-13-amd64 #1 SMP PREEMPT_DYNAMIC Debian 6.1.55-1 (2023-09-29) x86_64 GNU/Linux
```

-----------------------------

## Details of the problem

when running /bin/mpiexec, i get the error

```
$ /bin/mpiexec -np 2 hostname
--------------------------------------------------------------------------
A prefix was supplied to mpiexec that only contained slashes.

This is a fatal error; mpiexec will now abort.  No processes were launched.
--------------------------------------------------------------------------
```

It works fine without the path to `mpiexec`

```
$ mpiexec -np 2 hostname
vesubie
vesubie
```

the same problem was reported in https://github.com/open-mpi/ompi/issues/5613 and marked as completed.

Thanks for your help !

**Comment by bgoglin** [comment]
For the record, on recent Debian systems:
```
$ ll /bin 
lrwxrwxrwx 1 root root 7 19 sept.  2022 /bin -> usr/bin/
```
#5613 was slightly similar because /bin pointed to /usr/bin instead.

Running /usr/bin/mpiexec works fine.

**Comment by rhc54** [comment]
Ouch! That feels like a Debian bug. I don't have any immediate idea on how one would resolve that as it turns an absolute path into a relative one before the path even comes to me.

**Comment by bgoglin** [comment]
Relative symlink targets are valid (and widely used, for instance for libopenpal.so points to libopenpal.so.40 which points to libopenpal.so.40.30.2). "man ln" says "a relative link is interpreted in relation to its parent directory."

**Comment by rhc54** [comment]
Understood - but like I said, I don't know what to do with it as it just looks like a relative path was given to me. I can look at silencing the "only slashes" error, but the command is like;y to fail even if I silence it

**Comment by bgoglin** [comment]
I don't understand. Where do you get this path that "just looks like a relative path was given to me" ? If it comes from readlink, you're supposed to check whether the output starts with "/". If not, prefix it with the directory of the source. Hence in our case it's not relative anymore.

That said, I don't see many call to readlink() in ompi/prrte/pmix. One that may be buggy is in mca_fs_base_get_parent_dir() but it doesn't seem related to this case.

I assume we're talking about prrte here since that's where the error message is. Where does prrte actually reads the contents of the "/bin" symlink when we call "/bin/mpiexec"?

**Comment by rhc54** [comment]
I think it is safe to say that neither of us understands what the other is talking about 😄 The message is coming from PRRTE as it tries to decipher a prefix to pass along to the daemons so they can find their libraries.

I would say let's just table this until someone actually has time to fix it. We won't be releasing a PRRTE bug fix release until late first quarter next year, so there is plenty of time to address it.

**Comment by nfurmento** [comment]
I was able to get around the problem by having "/usr/bin" before "/bin" in my PATH. Maybe before fixing the bug you could modify the error message to inform users to not use a simlink path for mpiexec. Otherwise many users may get puzzled by the bug.

**Comment by rhc54** [comment]
I'm afraid that any change made to the error message would still not get released until March of next year. Besides, I'm skeptical that this symlink discussion is just a red herring - I suspect the problem is something basic in the prefix parsing code and has nothing to do with the symlink. Someone just needs to find a little time to chase it down, and time is scarce at the moment.

**Comment by rhc54** [comment]
I spent a few minutes trying to reproduce this issue as just a quick triage, and found:

* I cannot get an error, even when I created a symlink like the one described
* The reported error message does not exist in OMPI v5

I then realized that this isn't a bug report about the recently released v5, but rather one about the old v4.1.4 - my bad for not fully reading the system description. Your best solution is to update to v4.1.6 or v5.0, or perhaps someone over here in OMPI-land can help (I don't go back past the v5.0 release).

**Comment by nfurmento** [comment]
we just tried with v4.1.6 (which is the version in debian unstable), the problem is still there.
i do not know who is responsible for upgrading the debian package of OpenMPI, if i find out, i will let the person know the package needs an upgrade

**Comment by amckinstry** [comment]
Thanks @nfurmento for pointing this out to me.
This is part of the "/usr merge" transition in Debian/Ubuntu. See https://wiki.debian.org/UsrMerge

In Debian i'm testing out 5.0.0, but our current unstable is 4.1.6.
We have a package "usr-merge" which turns /bin and /lib into symlinks to /usr/bin. This is not universally installed (see the wiki page for details).

**Comment by jsquyres** [comment]
I don't think that the issue is the symbolic link.

@rhc54 and I are discussing this issue in higher bandwidth (i.e., Slack + webex).  Give us a little time to figure this one out.

**Comment by jsquyres** [comment]
The issue was unrelated to the sym link; it was more about `mpirun` being located in `/bin` (i.e., with an installation prefix of `/` vs. something longer, such as `/usr`, or `/opt/openmpi`, or somesuch).

A fix was committed upstream to PRRTE (because this affects Open MPI v5.0.0, too), and was back-ported to ORTE in Open MPI v4.1.x in https://github.com/open-mpi/ompi/pull/12093.

**Comment by nfurmento** [comment]
Thanks a lot Jeff!

**Comment by jsquyres** [comment]
This has been fixed in `main`, the `v4.0.x` branch, and the upstream PRRTE repo.  Still waiting for the fix to be pulled (via submodule update) to the `v5.0.x` branch.

**Comment by wenduwan** [comment]
This should have been fixed in latest 5.0.x release. @nfurmento could you please confirm?

**Comment by nfurmento** [comment]
works for me. Thanks a lot.
